### PR TITLE
Help fullscreen optimizations work better

### PR DESCRIPTION
**Pull request by gendlin**

### CODE REVIEW

File: src/i_video.c [hunk 1]
```diff
@@ -1616,6 +1615,10 @@ static void I_InitGraphicsMode(void)
             flags |= SDL_WINDOW_FULLSCREEN_DESKTOP;
         }
     }
+    else // !fullscreen
+    {
+        flags |= SDL_WINDOW_RESIZABLE;
```
Comment (fabiangreffrath) [inline src/i_video.c:1620]: Does `SDL_RenderSetLogicalSize()` work at all if this flag is not set?

Comment (gendlin) [reply src/i_video.c:1620]: Yes, seems to work fine.

File: src/i_video.c [hunk 2]
```diff
@@ -1581,7 +1581,6 @@ static void I_InitGraphicsMode(void)
     uint32_t flags = 0;
 
     // [FG] window flags
-    flags |= SDL_WINDOW_RESIZABLE;
```
Comment (ceski-1) [inline src/i_video.c:1584]: Something like this is needed to allow the window to be resized when toggling between fullscreen (borderless) and windowed:

```diff
diff --git a/src/i_video.c b/src/i_video.c
index 56a721b5..888a8217 100644
--- a/src/i_video.c
+++ b/src/i_video.c
@@ -397,6 +397,7 @@ static void I_ToggleFullScreen(void)
     {
         flags |= SDL_WINDOW_FULLSCREEN_DESKTOP;
         SDL_SetWindowGrab(screen, SDL_TRUE);
+        SDL_SetWindowResizable(screen, SDL_FALSE);
     }
 
     SDL_SetWindowFullscreen(screen, flags);
@@ -408,6 +409,7 @@ static void I_ToggleFullScreen(void)
     {
         SDL_SetWindowGrab(screen, SDL_FALSE);
         AdjustWindowSize();
+        SDL_SetWindowResizable(screen, SDL_TRUE);
         SDL_SetWindowSize(screen, window_width, window_height);
     }
 }
```

Comment (gendlin) [reply src/i_video.c:1584]: Added that, thanks.

Related to borderless mode, on this branch do you notice any differences in responsiveness when starting in borderless mode, then alt-entering out of then back into fullscreen? I do on my Win10 machine (with Game Bar and other overlays disabled). This is fixed if I change `I_InitGraphicsMode` to create the window as non-fullscreen and set `toggle_fullscreen=true` to switch it to fullscreen mode in the first update.

File: src/i_video.c [hunk 3]
```diff
@@ -1581,7 +1583,6 @@ static void I_InitGraphicsMode(void)
...
```
Comment (ceski-1) [inline src/i_video.c:1584]: Everything looks fine to me, but how does this help fullscreen optimizations work better?

Comment (gendlin) [reply src/i_video.c:1584]: Windows 10 seems to turn off FSO if the window style has `WS_THICKFRAME` or `WS_CAPTION`. The only reference I could find for this is an [old KB article](https://web.archive.org/web/20070219210133/http://support.microsoft.com/kb/179363/):

> b.	the window rectangle covers the entire monitor and the window does not have the WS_THICKFRAME style or does not have the WS_CAPTION style.

This seems to imply that Windows doesn't consider a window the size of the screen to be completely covering the screen (one of the conditions of FSO) if either of those two styles are applied.

Also, I had to turn of "Game Bar" in the Win10 settings (it helps to turn off both settings) to notice this, as that also seems to disable FSO for me. Disabling FSO in the executable compatibility settings has the same effect as reverting the changes in this branch, for me.

Comment (ceski-1) [reply src/i_video.c:1584]: > Windows 10 seems to turn off FSO if the window style has `WS_THICKFRAME` or `WS_CAPTION`.

With exclusive fullscreen, those `WS_*` values aren't set: https://github.com/libsdl-org/SDL/blob/SDL2/src/video/windows/SDL_windowswindow.c#L81 With non-exclusive fullscreen, it seems they are. If FSO isn't working, that would mean the present mode isn't one of the flip types, which can be checked with PresentMon. What is the present mode you're seeing before and after this PR?

Comment (gendlin) [reply src/i_video.c:1584]: > With exclusive fullscreen, those `WS_*` values aren't set

Yeah they are, `SDL_WINDOW_FULLSCREEN` is masked out here by `CREATE_FLAGS` before that other code gets called: https://github.com/libsdl-org/SDL/blob/SDL2/src/video/SDL_video.c#L1790

> If FSO isn't working, that would mean the present mode isn't one of the flip types

Not sure why - any references for this? As far as I know FSO is a complete black box and there's no way to introspect whether it's active or not for a particular window at any given time.

> What is the present mode you're seeing before and after this PR?

Hardware: Independent Flip

Comment (gendlin) [reply src/i_video.c:1584]: Side note: with these changes running in normal priority no longer causes massive stuttering when other things are going on in the background, like youtube.

Comment (ceski-1) [reply src/i_video.c:1584]: > Yeah they are, `SDL_WINDOW_FULLSCREEN` is masked out here by `CREATE_FLAGS` before that other code gets called: https://github.com/libsdl-org/SDL/blob/SDL2/src/video/SDL_video.c#L1790

Take a look at `SDL_FinishWindowCreation()` a few lines down. Or add this to the end of `I_InitGraphicsMode()` and look at the `-verbose` output:

```c
uint32_t testflags = SDL_GetWindowFlags(screen);
I_Printf(VB_DEBUG, "SDL_WINDOW_RESIZABLE: %s", (testflags & SDL_WINDOW_RESIZABLE) ? "true" : "false");
DWORD style = GetWindowLong(GetActiveWindow(), GWL_STYLE);
I_Printf(VB_DEBUG, "WS_THICKFRAME: %s", (style & WS_THICKFRAME) ? "true" : "false");
I_Printf(VB_DEBUG, "WS_CAPTION: %s", (style & WS_CAPTION) ? "true" : "false");
```

> Not sure why - any references for this? As far as I know FSO is a complete black box and there's no way to introspect whether it's active or not for a particular window at any given time.

Check out the links [here](https://github.com/fabiangreffrath/woof/pull/1754#issuecomment-2183163202) and [this](https://devblogs.microsoft.com/directx/demystifying-full-screen-optimizations/). You'll see "Legacy ..." or "Copy ..." when FSO benefits don't apply.

> Hardware: Independent Flip

So there was no change and you're optimized either way.

> Side note: with these changes running in normal priority no longer causes massive stuttering when other things are going on in the background, like youtube.

That's good for your specific configuration but I don't see any difference on mine. I don't mind these changes, but I'm still not convinced that it helps fullscreen optimizations, or that there's any reason to change anything.

Comment (gendlin) [reply src/i_video.c:1584]: > Take a look at `SDL_FinishWindowCreation()` a few lines down.

I'm looking..

> Or add this to the end of I_InitGraphicsMode() and look at the -verbose output:

Yes, the flag gets cleared later, but it seems like the damage is already done by the call to `CreateWindow` with `WS_THICKFRAME`. To test this, I confirmed that I don't see benefits here if I start the game in windowed mode then switch to borderless fullscreen, unless I remove the `SDL_WINDOW_RESIZABLE` altogether, or if I add the resizing handles manually after the `SDL_CreateWindow` call.

> Check out the links https://github.com/fabiangreffrath/woof/pull/1754#issuecomment-2183163202 and [this](https://devblogs.microsoft.com/directx/demystifying-full-screen-optimizations/).

I've seen that article but I'm not convinced FSO necessarily always involves a presentation mode change - there are also kernel scheduler changes that may be independently implemented of the swap chain stuff. I wish there was more insight into this via some API call but there doesn't appear to be one.

(Off-topic but related to the comment you linked above, I have a local SDL build with the d3d11 backend changed to use a flip-discard swap chain with the presentation flag to allow tearing - it plays very nice.)

> or that there's any reason to change anything

To me, it seems like we shouldn't be creating fullscreen windows with resizing handles.

Comment (ceski-1) [reply src/i_video.c:1584]: > Yes, the flag gets cleared later, but it seems like the damage is already done by the call to `CreateWindow` with `WS_THICKFRAME`. To test this, I confirmed that I don't see benefits here if I start the game in windowed mode then switch to borderless fullscreen, unless I remove the `SDL_WINDOW_RESIZABLE` altogether, or if I add the resizing handles manually after the `SDL_CreateWindow` call.

What damage and what lack of benefits?

Comment (ceski-1) [reply src/i_video.c:1584]: > To test this, I confirmed that I don't see benefits here if I start the game in windowed mode then switch to borderless fullscreen

Put the above code at the end of `I_ToggleFullScreen()` if you want to see window to borderless fullscreen transitions.

Starting with a window:
```
SDL_WINDOW_RESIZABLE: true
WS_THICKFRAME: true
WS_CAPTION: true
```
Then alt+enter:
```
SDL_WINDOW_RESIZABLE: true
WS_THICKFRAME: false
WS_CAPTION: false
```

File: src/i_video.c [hunk 4]
```diff
@@ -1635,6 +1636,8 @@ static void I_InitGraphicsMode(void)
 
     I_InitWindowIcon();
 
+    SDL_SetWindowResizable(screen, SDL_TRUE);
```
Comment (ceski-1) [inline src/i_video.c:1639]: What is the reasoning for this change?

Comment (gendlin) [reply src/i_video.c:1639]: To cover this case:

> start the game in windowed mode then switch to borderless fullscreen

I'll shift the code around to make it clearer we're not setting the fullscreen window as resizable.

**Review by ceski-1 — APPROVED** [review]
> To me, it seems like we shouldn't be creating fullscreen windows with resizing handles.

SDL has checks in some places, but not in others, to prevent this from happening so these changes seem reasonable to me.

**Pull request merged**